Texture.py

```diff
@@ -1,163 +1,326 @@
+        return (bits >> offset) & mask
+
+    width = size[0]
+    height = size[1]
+
+    if len(channel_sizes) == 3:
+        mode = "RGB"
+    elif len(channel_sizes) == 4:
+        mode = "RGBA"
+    else:
+        raise Exception("Unsupported channel_sizes %d" % len(channel_sizes))
+
+    img = Image.new(mode, (width, height))
+
+    # TODO: Is unswizzling actually necessary if textures are read via AGP?
+    # Need to set up a swizzled test case and verify behavior.
+
+    # FIXME: Unswizzle data on the fly instead
+    if swizzled:
+        data = nv2a.Unswizzle(data, bits_per_pixel, (width, height), pitch)
+
+    pixels = img.load()  # create the pixel map
+
+    assert bits_per_pixel % 8 == 0
+
+    for y in range(height):
+        for x in range(width):
+
+            pixel_offset = y * pitch + x * bits_per_pixel // 8
+            pixel_bytes = data[pixel_offset : pixel_offset + bits_per_pixel // 8]
+            pixel_bits = int.from_bytes(pixel_bytes, byteorder="little")
+
+            pixel_channels = ()
+            for channel_offset, channel_size in zip(channel_offsets, channel_sizes):
+                channel_value = get_bits(pixel_bits, channel_offset, channel_size)
+
+                # Normalize channel
+                if channel_size > 0:
+                    channel_value /= (1 << channel_size) - 1
+                    channel_value = int(channel_value * 0xFF)
+                else:
+                    channel_value = 0x00
+
+                pixel_channels += (channel_value,)
+            pixels[x, y] = pixel_channels
+
+    return img
+
+
+def surface_color_format_to_texture_format(fmt, swizzled):
+    """Convert nv2a draw format to the equivalent Texture format."""
+    if fmt == 0x3:  # ARGB1555
+        return 0x3 if swizzled else 0x1C
+
+    if fmt == 0x5:  # RGB565
+        return 0x5 if swizzled else 0x11
+
+    if fmt in [0x7, 0x08]:  # XRGB8888
+        return 0x7 if swizzled else 0x1E
+
+    if fmt == 0xC:  # ARGB8888
+        return 0x6 if swizzled else 0x12
+
+    raise Exception(
+        "Unknown color fmt %d (0x%X) %s"
+        % (fmt, fmt, "swizzled" if swizzled else "unswizzled")
+    )
+
+
+def surface_zeta_format_to_texture_format(fmt, swizzled, is_float):
+    """Convert nv2a zeta format to the equivalent Texture format."""
+    if fmt == 0x1:  # Z16
+        if is_float:
+            return 0x2D if swizzled else 0x31
+        return 0x2C if swizzled else 0x30
+
+    if fmt == 0x2:  # Z24S8
+        if is_float:
+            return 0x2B if swizzled else 0x2F
+        return 0x2A if swizzled else 0x2E
+
+    raise Exception(
+        "Unknown zeta fmt %d (0x%X) %s %s"
+        % (
+            fmt,
+            fmt,
+            "float" if is_float else "fixed",
+            "swizzled" if swizzled else "unswizzled",
+        )
+    )
+
+
+def read_texture_parameters(xbox: Xbox) -> TextureParameters:
+    """Reads the current texture state"""
+    color_pitch = xbox.read_u32(0xFD400858)
```

Comment:
Would like to these addresses eventually replaced with named constants. Can wait for follow up however